src/wp-settings.php

```diff
@@ -389,6 +389,11 @@
 require ABSPATH . WPINC . '/interactivity-api/class-wp-interactivity-api-directives-processor.php';
 require ABSPATH . WPINC . '/interactivity-api/interactivity-api.php';
 
+// Register the default theme directory root.
+// This has to happen before wp_script_modules is registered
+// because wp_script_modules needs to know if the current theme is a block theme.
```

Comment:
Do you know why `wp_script_modules` need to know about the theme directory?

Reply:
But printing header/footer happen later now? (I mean later than the original place of this call)

Reply:
It's not clear from [the original commit](https://github.com/WordPress/wordpress-develop/commit/03cb3e30f20b1cdff0c4b0fb3bde2bcd3c4a1c30), why the `add_hooks()` call was added here.

Reply:
Well, it's not happening _later_. If anything the hooks are added earlier now, right after the theme is registered.

But the _printing_ of these modules is still happening at the same priority of these hooks.

Reply:
What I mean is that instead of doing the `wp_is_block_theme()`, run the printing hook in both footer and header and check the block theme in the callback instead. At that point, we're certain that the directory is correct. This avoids any race condition.

I guess for me changing order of calls can be problematic sometimes, so this was the safe option but looking at what the add_hooks function does, it probably safer to move it later anyway.

Reply:
I think moving the `wp_is_block_theme` inside the footer and header hooks would make the code harder to reason about since the two code forks would no longer be co-located, so the relationship between them would be unclear.

Reply:
> I think moving the wp_is_block_theme inside the footer and header hooks would make the code harder to reason about since the two code forks would no longer be co-located

Maybe not necessarily not co-located, but a bit duplicated logic, as we'd have to add checks like this in three methods:

```php
if ( wp_is_block_theme() && 'wp_footer' === current_action() ) {
	// Only load in the head for block themes.
	return;
} elseif ( ! wp_is_block_theme() && 'wp_head' === current_action() ) {
	// Only load in the footer for classic themes.
	return;
}
```

Reply:
I agree it's a bit duplicated but the truth is actually we don't need to know that before "init", so it's less fragile. 

I'm ok with the current fix though.

Reply:
cc @luisherranz @westonruter for thoughts

Reply:
I think `add_hooks` is being called too early here still. The Customizer, for example, can dynamically change the theme later at the `setup_theme` action. Same could be done by other plugins, like those which serve a mobile theme.

Therefore, Instead of:

```php
wp_script_modules()->add_hooks();
wp_interactivity()->add_hooks();
```

I suggest:

```php
add_action( 'after_setup_theme', array( wp_script_modules(), 'add_hooks' ) );
add_action( 'after_setup_theme', array( wp_interactivity(), 'add_hooks' ) );
```

Also, the code could be removed from `src/wp-settings.php` and instead located in a more natural place: `src/wp-includes/default-filters.php`

It could go after these lines:

https://github.com/WordPress/wordpress-develop/blob/1d924aefe955e04b6bd3af0baafaeb961ddf2b97/src/wp-includes/default-filters.php#L563-L565

Reply:
`default-filters.php` is loaded way before `wp_script_modules()` and `wp_interactivity()` are loaded, so that won't work.

But using `after_setup_theme` sounds reasonable.